Angular v13 features and examples (#956)
oktadeveloper/okta-blog: _source/_posts/2021-11-10-angular-v13.md

---
layout: blog_post
title: "What You Need to Know about Angular v13"
author: alisa-duncan
by: advocate
communities: [javascript]
description: "Highlights of a few Angular v13 features with code snippets"
tags: [javascript, typescript, angular]
tweets:
- "Excited for @angular v13? 🎉 Check out some of the new features with code demos! #angular"
- "Check out some use cases for @angular v13 features. What's your favorite feature? #angular"
- ""
image: blog/angular-v13/angular-v13-social.jpg
type: awareness
---

> Angular v13 arrived! And with it comes a lot of exciting new features and updates. Angular continues to improve run-time performance, decrease compilation time, promote good software development practices, enhance developer experience, and keep up to date with dependencies such as TypeScript and RxJS. Is anyone else excited about RxJS v7?! 🤩 Let's take a look at a few of the many new exciting features in Angular v13 with some code examples using authentication. 

Reviewer: ```suggestion
Angular v13 has arrived! And with it come a lot of exciting new features and updates. Angular continues to improve runtime performance, decrease compilation time, promote good software development practices, enhance developer experience, and keep up to date with dependencies such as TypeScript and RxJS. Is anyone else excited about RxJS v7?! 🤩 Let's take a look at a few of the many new exciting features in Angular v13 with some code examples using authentication. 
```
grammar and verb tense

{% include toc.md %}

## So long, View Engine!

> With v13, View Engine is no longer available and replaced by Ivy. Ivy is the future of Angular applications, and the benefits are tremendous. After initially landing in v12, Ivy paves the way for new feature updates, simplifies existing capabilities, and increases rendering performance.

Reviewer: ```suggestion
With v13, View Engine is no longer available, replaced by Ivy. Ivy is the future of Angular applications, and the benefits are tremendous. After initially landing in v12, Ivy paves the way for new feature updates, simplifies existing capabilities, and increases rendering performance.
```
grammar nit

> An example of simplifying existing capabilities is how to create components dynamically. Dynamic components are components loaded at runtime. You might use this when you need to tailor visible components based on responses from a server call. Previously, to dynamically create components, you had to use the `ComponentFactoryResolver`. Now the API takes the Component type directly. Let's look at how this works.

Reviewer: ```suggestion
The way you can now create components dynamically is an example of how this release simplifies existing capabilities. Dynamic components are components loaded at runtime. You might use this when you need to tailor visible components based on responses from a server call. Previously, to dynamically create components, you had to use the `ComponentFactoryResolver`. Now the API takes the Component type directly. Let's look at how this works.
```
making the opening sentence more declarative.

### Set up the Authorization Server 

In my example, I am authenticating using OAuth 2.0 with OpenID Connect through Okta. In my Okta Authorization Server, I added a custom claim named `pet_preference` to my identity token. Does this sound only vaguely familiar? Refresh your understanding of OAuth 2.0 and OpenID Connect by reading "[The Things to Keep in Mind about Auth](https://developer.okta.com/blog/2021/10/29/things-to-keep-in-mind-about-auth)".

With this custom claim, I want to display dynamic components specific to the value. The app uses the claim value to rotate through images of pets that match their preference.

### Prepare the components

To write something like this, you'll first create components for the customized views of different pets that all implement the same interface, such as one named `PreferenceComponentInterface`. The interface has a property named `data` so you can pass in customized information. You'll also need to get the `pet_preference` custom claim value and retrieve content specifically for it. Once you have the data for the preferred pet, you can add the code to create dynamic components.

### Implement the dynamic component creation

> Create a `Directive` to house the view

Reviewer: ```suggestion
Create a `Directive` to house the view.
```
punctuation

```ts
import { Directive, ViewContainerRef } from '@angular/core';

@Directive({
  selector: '[appPreference]'
})

export class PreferenceDirective {
  constructor(public viewContainerRef: ViewContainerRef) { }
}
```

> And a container component that handles the dynamic creation of the component. I simplified the code snipped to focus on the v13 change. Stay tuned for an upcoming post on a walkthrough to create this application.

Reviewer: ```suggestion
And a container component that handles the dynamic creation of the component. I simplified the code snipped to focus on the v13 change. In an upcoming post, I'll do a complete walkthrough of creating this application.
```
flow

```ts
class PreferenceItem {
  component: Type<PreferenceComponentInterface>;
  info: any;
}

@Component({
  selector: 'app-secret-preference',
  template: `<ng-template appPreference></ng-template>`
 })

export class SecretPreferenceComponent implements OnInit {
  @ViewChild(PreferenceDirective, {static: true}) private host!: PreferenceDirective;
  
  private preferences: PreferenceItem[] = [{
    component: MyCustomComponent,
    info: 'example'
  }, {
    component: MyOtherCustomComponent,
    info: {
      isTrue: true
    }
  }];
  
  public ngOnInit(): void {
    const preferredItem = this.preferences[0];
    const compRef = this.host.viewContainerRef.createComponent<PreferenceComponentInterface>(preferredItem);
    compRef.instance.data = preferredItem.info;
  }
}
```

> Notice you didn't have to inject the `ComponentFactoryResolver` to do this. The API is more straightforward to use. A more realistic use-case might be a custom claim for the department and displaying customized content based on the authenticated user's department.

Reviewer: ```suggestion
Notice you didn't have to inject the `ComponentFactoryResolver` to do this. The API is more straightforward to use. A more realistic use case might be a custom claim for the department and displaying customized content based on the authenticated user's department.
```
no hyphen needed

## Clean-up after each test

> Testing Angular apps is much improved with better cleanup of the DOM between tests defaulted in v13. Now, Angular does a better job of cleaning up the DOM between tests, and tests run faster and more performantly with better isolation between tests.

Reviewer: ```suggestion
Testing Angular apps is much improved in v13, and better cleanup of the DOM between tests is the new default. Now, Angular does a better job of cleaning up the DOM, and tests run faster and more performantly with better isolation between tests.
```
clarifying

      declarations: [ ProfileComponent ],
      providers: [
        { provide: OktaAuthStateService, useValue: authStateSpy }
      ],
      teardown: { destroyAfterEach: false } // flakey tests!
    });
  });

  it('should show text that welcomes the user by name', () => {
    fixture = TestBed.createComponent(ProfileComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
      
    const spanEl = fixture.debugElement.query(By.css('span'));
    expect(spanEl).toBeTruthy();
    expect(spanEl.nativeElement.innerHTML).toEqual('Welcome Test Name!');
  });

  it('should show text that welcomes a user with no name', () => {
    (Object.getOwnPropertyDescriptor(authStateSpy, 'authState$')?.get as jasmine.Spy).and.returnValue(of({
      isAuthenticated: true,
      idToken: {...idToken, claims: {sub: 'sub'}}
    }));
      
    fixture = TestBed.createComponent(ProfileComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    const spanEl = fixture.debugElement.query(By.css('span'));
    expect(spanEl).toBeTruthy();
    expect(spanEl.nativeElement.innerHTML).toEqual('Welcome stranger!');
  });
});
```

> We can configure the teardown cleanup in this test suit by changing the value of `destroyAfterEach`. With this change enabled by default, you get the best option of writing new tests with better performance and disabling it for tests that aren't ready to consume the changes. Angular has our backs on non-breaking updates!

Reviewer: ```suggestion
We can configure the teardown cleanup in this test suite by changing the value of `destroyAfterEach`. With this change enabled by default, you have the option of writing new tests with better performance. Plus, you can disable the default for tests that aren't ready to consume the changes. Angular has our backs on non-breaking updates!
```
clarifying

## Enhanced developer tooling

> Creating Angular apps is now easier with enhancements to developer tooling and compilation time improvements.

Reviewer: ```suggestion
Creating Angular apps is now easier with developer tooling enhancements and compilation time improvements.
```
parallel structure in sentence to ease readability

The first thing you might notice is how much faster the build time is. A lot of work went into improving the build time. By adding [a build cache](https://angular.io/cli/cache), the Angular team found build time improvements of **up to 65+%**!

The Angular Language Service's tooling improvement to auto-apply optional chaining feels like a small thing but is such an appreciated feature. I can focus on the object property I want to use, and the tooling handles the optional chaining for me!

{% img blog/angular-v13/auto-apply-optional-chaining.gif alt:"Auto applying optional chaining in VS Code for a deep object with optional properties" width:"600" %}{: .center-image }

> One thing to note about the demo shown above is that it is for demonstration purposes and exaggerates capabilities to showcase the new feature, so please remember to apply sound software development practices to your applications.

Reviewer: ```suggestion
Note: The example demo shown above is for demonstration purposes, and exaggerates capabilities to showcase the new feature. Please remember to apply sound software development practices to your real-world applications.
```
readability

 
## So long, Internet Explorer!

> In this release, we also say goodbye to Internet Explorer. [Dropping Internet Explorer support](https://github.com/angular/angular/issues/41840) means lots of doors open up for new features, refactoring the code-base, and keeping up with security best practices. With extra polyfills, you can use Okta's Auth JS library in Internet Explorer, but the recommendation is to migrate away from it.

Reviewer: ```suggestion
In this release, we also say goodbye to Internet Explorer. [Dropping Internet Explorer support](https://github.com/angular/angular/issues/41840) means lots of doors open up for new features, refactoring the code base, and keeping up with security best practices. With extra polyfills, you can use Okta's Auth JS library in Internet Explorer, but the recommendation is to migrate away from it.
```
code base: two word spelling, no hyphen

## All the Angular v13 details

> These are the features that I found most interesting, but there's a lot more in Angular v13! You can read more about them in the [Angular team's blog post](https://blog.angular.io/angular-v13-is-now-available-cce66f7bc296). You can read about the changes in the new version of [TypeScript](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/) and [RxJS](https://rxjs.dev/6-to-7-change-summary).

Reviewer: ```suggestion
These are the features that I found most interesting, but there are lots more in Angular v13! You can read more about them in the [Angular team's blog post](https://blog.angular.io/angular-v13-is-now-available-cce66f7bc296). You can read about the changes in the new version of [TypeScript](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/) and [RxJS](https://rxjs.dev/6-to-7-change-summary).
```
subject-verb agreement (lots<-> them)

## Upcoming changes

> Angular keeps improving and growing. The [subsequent releases](https://angular.io/guide/roadmap) should prove exciting, with features such as strict typing for forms and stand-alone components in the works.

Reviewer: ```suggestion
Angular keeps improving and growing. The [subsequent releases](https://angular.io/guide/roadmap) should prove exciting, with features such as strict typing for forms and standalone components in the works.
```
spelling/terminology usage
